// @aztec/bb.js/dest/node-cjs/crs/node/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Crs = void 0;
const tslib_1 = require("tslib");
const net_crs_js_1 = require("../net_crs.js");
const ignition_files_crs_js_1 = require("./ignition_files_crs.js");
const fs_1 = require("fs");
const promises_1 = require("fs/promises");
const debug_1 = tslib_1.__importDefault(require("debug"));
const debug = (0, debug_1.default)('bb.js:crs');
/**
 * Generic CRS finder utility class.
 */
class Crs {
    constructor(numPoints, path) {
        this.numPoints = numPoints;
        this.path = path;
    }
    static async new(numPoints, crsPath = './crs') {
        const crs = new Crs(numPoints, crsPath);
        await crs.init();
        return crs;
    }
    async init() {
        (0, fs_1.mkdirSync)(this.path, { recursive: true });
        const size = await (0, promises_1.readFile)(this.path + '/size', 'ascii').catch(() => undefined);
        if (size && +size >= this.numPoints) {
            debug(`using cached crs of size: ${size}`);
            return;
        }
        const crs = ignition_files_crs_js_1.IgnitionFilesCrs.defaultExists() ? new ignition_files_crs_js_1.IgnitionFilesCrs(this.numPoints) : new net_crs_js_1.NetCrs(this.numPoints);
        if (crs instanceof net_crs_js_1.NetCrs) {
            debug(`downloading crs of size: ${this.numPoints}`);
        }
        else {
            debug(`loading igntion file crs of size: ${this.numPoints}`);
        }
        await crs.init();
        (0, fs_1.writeFileSync)(this.path + '/size', this.numPoints.toString());
        (0, fs_1.writeFileSync)(this.path + '/g1.dat', crs.getG1Data());
        (0, fs_1.writeFileSync)(this.path + '/g2.dat', crs.getG2Data());
    }
    /**
     * G1 points data for prover key.
     * @returns The points data.
     */
    getG1Data() {
        return (0, fs_1.readFileSync)(this.path + '/g1.dat');
    }
    /**
     * G2 points data for verification key.
     * @returns The points data.
     */
    getG2Data() {
        return (0, fs_1.readFileSync)(this.path + '/g2.dat');
    }
}
exports.Crs = Crs;
